tests/api.c

```diff
@@ -58953,10 +58958,94 @@ static int test_ticket_nonce_read_cb(WOLFSSL *ssl,
 
     return read_sz;
 }
+#endif /* WOLFSSL_MEMIO_CBS */
+
+#if defined(WOLFSSL_DTLS) && !defined(WOLFSSL_NO_TLS12) &&                     \
+    !defined(NO_WOLFSSL_CLIENT) && !defined(NO_WOLFSSL_SERVER) &&              \
+    !defined(NO_OLD_TLS)
+static int test_WOLFSSL_dtls_version_alert(void)
+{
+    struct test_memio_ctx test_ctx = { 0 };
+    WOLFSSL_CTX *ctx_c, *ctx_s;
+    WOLFSSL *ssl_c, *ssl_s;
+    int ret;
+
+    ctx_c = wolfSSL_CTX_new(wolfDTLSv1_2_client_method());
+    ctx_s = wolfSSL_CTX_new(wolfDTLSv1_server_method());
+
+    wolfSSL_CTX_set_verify(ctx_s, WOLFSSL_VERIFY_NONE, 0);
+    wolfSSL_CTX_set_verify(ctx_c, WOLFSSL_VERIFY_NONE, 0);
+    ret = wolfSSL_CTX_use_PrivateKey_file(ctx_s, svrKeyFile,
+        WOLFSSL_FILETYPE_PEM);
+    if (ret != WOLFSSL_SUCCESS)
+        return- -1;
+    ret = wolfSSL_CTX_use_certificate_file(ctx_s, svrCertFile,
+                                           WOLFSSL_FILETYPE_PEM);
+    if (ret != WOLFSSL_SUCCESS)
+        return -1;
+    wolfSSL_SetIORecv(ctx_c, test_memio_read_cb);
+    wolfSSL_SetIOSend(ctx_c, test_memio_write_cb);
+    wolfSSL_SetIORecv(ctx_s, test_memio_read_cb);
+    wolfSSL_SetIOSend(ctx_s, test_memio_write_cb);
+    ssl_c = wolfSSL_new(ctx_c);
+    ssl_s = wolfSSL_new(ctx_s);
+    if (ssl_c == NULL || ssl_s == NULL)
+        return -1;
+    wolfSSL_SetIOWriteCtx(ssl_c, &test_ctx);
+    wolfSSL_SetIOReadCtx(ssl_c, &test_ctx);
+    wolfSSL_SetIOWriteCtx(ssl_s, &test_ctx);
+    wolfSSL_SetIOReadCtx(ssl_s, &test_ctx);
+
+    /* client hello */
+    ret = wolfSSL_connect(ssl_c);
+    if (ret == 0 || ssl_c->error != WANT_READ )
+        return -2;
+    /* hrr */
+    ret = wolfSSL_accept(ssl_s);
+    if (ret == 0 || ssl_s->error != WANT_READ )
+        return -2;
+    /* client hello 1 */
+    ret = wolfSSL_connect(ssl_c);
+    if (ret == 0 || ssl_c->error != WANT_READ )
+        return -2;
+    /* server hello */
+    ret = wolfSSL_accept(ssl_s);
+    if (ret == 0 || ssl_s->error != WANT_READ )
+        return -2;
+    /* should fail */
+    ret = wolfSSL_connect(ssl_c);
+    if (ret == 0 || ssl_c->error != VERSION_ERROR)
+        return -3;
+    /* shuould fail */
+    ret = wolfSSL_accept(ssl_s);
+    if (ret == 0 ||
+        (ssl_s->error != VERSION_ERROR && ssl_s->error != FATAL_ERROR))
+        return -3;
+
+    wolfSSL_free(ssl_c);
+    wolfSSL_free(ssl_s);
+    wolfSSL_CTX_free(ctx_c);
+    wolfSSL_CTX_free(ctx_s);
```

Comment:
Since this is used twice (and will probably be used more in the future) can you please refactor this section into an easy to use and invoke function? Something like the `test_wolfSSL_client_server_nofail` API would be nice.

Reply:
I refactored the boilerplate setup code